### ISSUE TITLE
Object type with function properties

### ISSUE THREAD

**Issue by jareware** [facebook/flow#273]
Not really sure what would be an appropriate title, please amend as you see fit.

Given this prelude:
```ts
/* @flow */ 
type SomethingDoer = {
  doSomething: (input: number) => string;
};
```
Exhibit A (passes):
```ts
function create() : SomethingDoer {
  return {
    doSomething() { // I would have expected this to conflict with SomethingDoer
      return '';
    }
  };
}
```
Exhibit B (fails):
```ts
function create() : SomethingDoer {
  return {
    doSomething(input : string) { // "This type is incompatible ..." as is expected
      return '';
    }
  };
}
```
Exhibit C (passes):
```ts
function create() : SomethingDoer {
  return {
    doSomething(input : boolean) { // I would have expected this to conflict with SomethingDoer
      return '';
    }
  };
}
```
Ideas? I'm on `0.4.0`.

**Comment by samwgoldman** [comment]
This can be expressed more simply:
```
type Contra = (x: number) => void;

var exA: Contra = function() {}
//var exB: Contra = function(x: string) {}
var exC: Contra = function(x: boolean) {}
```

I'm not an expert, so I'm probably not going to use the right words to explain this. I would appreciate it if someone would help correct my wording.

Example A follows because a function taking no arguments is a subtype of a function taking any arguments. This has a straightforward intuition, the function taking no arguments can't make any invalid assumptions—it can't make any assumptions at all.

Example C follows from the subtyping relationship `number < boolean` and the contravariance of function arguments. I believe the subtyping relationship exists so that numbers can be used in constructs like `if`.

In other words, this is the expected behavior.

**Comment by jareware** [comment]
Oh, thanks a lot. I totally forgot to consider the subtype relationships with my examples, and also managed to pick a particularly hairy Example C (as booleans are indeed complicated by the truthiness/falsyness of JavaScript).

It's quite clear now, sorry for the noise! :)
